Examples/ResourceLeak/README.md

```diff
@@ -0,0 +1,28 @@
+## Procedure-local Bugs of Resource Leaks
+
+A simple example of **procedure-local bug of resource leak** is a resource leak in the context of incomplete code (e.g. without main) with inter-procedural reasoning. In this case, the usual notion that all resource allocated must be eventually freed does not apply in the context of incomplete code. **CSharpCodeAnalyzer** is designed to detect this kind of resource leaks by applying the following principle: **When a new object is allocated during the execution of a procedure, it is the procedure’s responsibility to either deallocate the object or make it available to its callers; there is no such obligation for objects received from the caller.** 
+
+Based on the nature of procedure-local bugs of resource leaks, **CSharpCodeAnalyzer** is capable of detecting resource leaks in the following scenarios that are unrelated to multi-threading:
+
+### 1. Standard idioms: 
+
+Some objects in C#, the resources, are supposed to be disposed when you stop using them, and failure to dispose is a resource leak. Resources include input streams, output streams, readers, writers, sockets, http connections, cursors, json parsers, etc.
+							
+### 2. Nested allocations: 
+When a resource allocation is included as an argument to a constructor, if the constructor fails it can leave an an unreachable resource that no one can dispose.
+	
+For example *gzipStream = new GZipStream(new FileStream(out, FileMode.Create), CompressionMode.Compress);* is bad in case the outer constructor, *GZipStream*, throws an exception. In that case, no one will have a hold of the *FileStream* and so no one will be able to dispose it.
+	
+### 3. Allocation of Cursor resources:
```

Comment:
I used the phrase "Resource Allocation inside libraries" as the title for section 3.  This phrase comes from Infer team (next paragraph is the original description of this case):

"Allocation of JSonParser and Cursor resources
**Some resources are created inside libraries instead of by "new".**

Cursor is an interface, the actual resources are something like SQLiteCursor. So, every time you call a function that returns a Cursor object, there is an allocation.

For instance, in the functions from SQLiteDatabase query(…) and rawQuery(…) allocate a cursor resource. For SQLiteQueryBuilder, ContentProviderClient, ContentResolver. MediaStore and DownloadManager it is only query(…) Cursor objects cursor created by these functions need to be closed (i.e., cursor.close()).

Similarly, JsonParser is an abstract class, and create a resource in functions from the class JsonFactory createParser(byte[] data) createParser(byte[] data, int offset, int len) createParser(String content) createParser(URL url) createParser(File f) JsonParser objects js created by these functions need to be closed (jp.close()). On the other hand . JasonParsers gotten from createParser(InputStream in) and createParser(Reader r) give you JsonParsers that don’t need to be closed. This is because they receive the resource from somewhere that will maintain the responsibility to close it.

"